### ISSUE TITLE
Implement ERC-7540

### ISSUE THREAD

**Issue by Amxx** [OpenZeppelin/openzeppelin-contracts#4761]
[ERC-7540](https://eips.ethereum.org/EIPS/eip-7540) is in the review phase, and looks like a good candidate for inclusion.

**Comment by mw2000** [comment]
@Amxx I'd love to help with this

**Comment by Amxx** [comment]
> I'm still unsure about how to approach overriding deposit, redeem, withdraw, and mint functions.

This is exactly why this issue needs to be discussed and researched BEFORE we make any PR. As the tag shows, this is this only an idea. We may not even do that at all. 

One thing we need to figure out is the process for going from `pendingDepositRequest` to `maxDeposit`. Because the redeem should `deposit` should consume the `maxDeposit`, NOT the `pendingDepositRequest`.

Please do not rush to propose implementations. The ERC is still in draft, and we need to hear from (multiple) potential users before we start designing an actual solution.

**Comment by 0xTimepunk** [comment]
Any updates on this?

**Comment by mw2000** [comment]
@Amxx @0xTimepunk more than happy to pickup where we left off on this PR

**Comment by alphastorm** [comment]
The ERC was finalized two days ago 🎉 

<img width="627" alt="Screenshot 2024-06-28 at 12 37 48 PM" src="https://github.com/OpenZeppelin/openzeppelin-contracts/assets/972382/58fae95b-5fdf-4ba2-bb47-38d6dc856ef1">

cc @Joeysantoro

**Comment by yorhodes** [comment]
@alphastorm @Joeysantoro @mw2000 
I am very interested in a reference implementation for this which composes with the rest of the OpenZeppelin ERC20/4626 lifecycle hooks

**Comment by kevin-pv01** [comment]
We'd be interested in using 7540 if there was an OZ reference implementation.
Do we know if any work on this is planned?

**Comment by TalCrypto** [comment]
Any update on this implementation?
I represent logarithm.fi and we are building an asynchronous redeem vault now.
Would be happy if there were some OZ references.

**Comment by zjesko** [comment]
same, would love to have a OZ reference (or any other) for buildling trading vaults

**Comment by mw2000** [comment]
@zjesko @yorhodes @kevin-pv01 
 
There's a reference implementation in this repo (thanks to @TalCrypto for mentioning)  -> https://github.com/ERC4626-Alliance/ERC-7540-Reference

Alternatively, I plan to sharpen up my old implementation in the #4776 PR, if there's demand for it. I would love to get input on what you guys want to see in a reference implementation, especially if the other implementation is not a good fit.

**Comment by hieronx** [comment]
Just wanted to share my thoughts here after developing the above mentioned reference implementation + our implementation at Centrifuge (https://github.com/centrifuge/liquidity-pools/blob/main/src/ERC7540Vault.sol).

I think there is a few points of discussion that should be decided upon first, before implementing the full OpenZeppelin version. I'll list what I think those are here:

### Vault classes support
You can categorize vaults in 4 classes:
A. Fully synchronous (`IERC4626`)
B. Asynchronous on deposit (`IERC7540Deposit`)
C. Asynchronous on redeem (`IERC7540Redeem`)
D. Fully asynchronous (`IERC7540Deposit` + `IERC7540Redeem`)

A is already supported by https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/extensions/ERC4626.sol

The main use case I see for C is Liquid Staking Tokens (LSTs).

Most RWA vaults (like Centrifuge) are likely D.

I would propose, similar to https://github.com/ERC4626-Alliance/ERC-7540-Reference/tree/main/src, to support all 4 classes through contract inheritance.

### Request fulfillment methods
I can roughly think of 3 groups of methods for fulfilling requests:
1. **Manual per user**: an admin controlled function needs to be called to fulfill a request for a specific user. This is the case for most RWAs.
2. **Batched**: there is some grouping of requests (e.g. based on a time window) and each batch of requests is fulfilled manually by an admin controlled function
3. **Time-based**: LSTs are a good example, where the redemptions are fulfilled with e.g. a 7 day delay

In my opinion, batched fulfillments will likely be highly custom per implementation, and hard to generalize through an OpenZeppelin reference implementation (beyond the basic building blocks of ERC4626)

I would propose implementing the base OZ implementation as 1, and allowing the fulfillment method to be `virtual` to support custom implementations.

This can also support 3, by overriding the fulfillment method to be permissionless if the time delay has passed.

### ERC-7575 support
ERC-7540 enforces that each implementation also supports ERC-7575. This can be simply done by adding
```solidity
function share() external view returns (address) { return address(this); }
```

to the contract.

I would propose adding this to the standard ERC-4626 contract. It is a helpful addition to any ERC-4626 implementation, that can support the multi-asset vault use cases in the future (potentially through a future OZ reference implementation building on ERC-4626). And ERC-7540 will inherit from ERC-4626 contracts thus then automatically inheriting the ERC-7575 support.

### Implementation

I actually had started porting over the reference implementation linked above to the OZ repository, see this branch: https://github.com/hieronx/openzeppelin-contracts/blob/erc7540/contracts/token/ERC20/extensions/ERC7540.sol I haven't yet had time to take this further yet. @mw2000 if you want to do this, happy to support you!

I would say the above questions should be agreed here in the thread first, before continuing with the implementation.

@zjesko @yorhodes @kevin-pv01 @TalCrypto @Amxx curious on your thoughts!

**Comment by TalCrypto** [comment]
I totally agree on your thoughts @hieronx.
Your classifications cover everything that is needed to implement ERC7540.
The fulfillment methods are really dynamic based on protocols' needs.
For example, we are building AsyncRedeemVault and the fulfillment method is batched one.
Due to this specific requirement, I can't find any reference on this.
We've already built it by adhering ERC4626, but it requires additional funcions beyond the standard to withdraw.
Anyways, I will try to adhere ERC7540, and then will share my implementation here.
Thank you for bringing up this topic here @mw2000

**Comment by mw2000** [comment]
@hieronx happy to take it over if it's still in flight :)

**Comment by hieronx** [comment]
Thanks all for the feedback!

I've pushed the setup for the proposal above here: https://github.com/OpenZeppelin/openzeppelin-contracts/pull/5294 It is based on the existing reference implementation, supports vault classes B, C and D separately through inheritance, and implements manual fulfilment per user.

@mw2000 if you want to take over the implementation, that would be great.

It would be ideal to get some feedback on the design of the ERC7540 implementation from the maintainers (cc @Amxx and anyone else) before too much time is invested in implementing one approach.

**Comment by hieronx** [comment]
@Amxx was wondering if you have any feedback on this design proposal? Would love to move forward on this as there is a lot of requests for a reference implementation for this standard.

**Comment by Amxx** [comment]
We didn't really got time to review that latelly. Our focus has been on very different stuff for DevCon and 5.2.

I'm adding that as a candidate for the next release.

**Comment by hieronx** [comment]
Sounds good! Happy to contribute in any way I can to this.

**Comment by hlauinfo** [comment]
Hey @amxx curious which implementation you'll be looking at for inclusion. Am I correct to assume that you're sticking with #5294 with the closure of #5457?

**Comment by Amxx** [comment]
Hello @hlauinfo.

So far no definite decision has been made regarding 7540. 

#5457 was obviously not allign with this library design principle, and it felt like the effort to "fix it" would be greater thatn the effort to start from scratch. Also it missed a lot of the design thinking that is necessary for such an intricate ERC.

Ideally I'd like to have the time to review #5294, and merge it in an upcomming version. But the truth is, we haven't got basically any request for that ERC to be implemented, and without any such request, it was never prioritised. I can bump it so it appears on our roadmap, but that won't guarantee that something deemed "more important" won't take over our bandwidth.

**Comment by GuiDupont** [comment]
Regarding this proposal, at Lagoon.finance we built our infrastructure on top of ERC7540. Having a reference implementation would have protected us against errors made on the interpretation of the standard. Asynchronicity has become a key element of DeFi nowadays. The method proposed seems appropriate.

**Comment by drinkius** [comment]
We are also researching ERC-7540 at UltraYield and it'd be great to rely on OZ implementation instead our custom built one, but the time is also of the essence so it's possible that we'll deploy prior to having a PR merged

**Comment by vikramarun** [comment]
Adding in support here from Superform on the direction taken in #5294 by @hieronx. Would be very helpful to have this before any large scale efforts to integrate across DeFi in the second half of the year

**Comment by alphastorm** [comment]
### 🚀 Cove is now live on mainnet!

We’ve made the core of the **Cove Protocol** source available at [`Storm-Labs-Inc/cove-contracts-core`](https://github.com/Storm-Labs-Inc/cove-contracts-core) — including **[`BasketToken.sol`](https://github.com/Storm-Labs-Inc/cove-contracts-core/blob/master/src/BasketToken.sol)**, our publicly released implementation of **ERC-7540** (shoutout to our advisor @Joeysantoro!).

- **Battle-tested security** – Three independent deep-dive audits completed pre-launch:  
  - ✅ **Zellic** – Dec 23 2024 ([report](https://github.com/Storm-Labs-Inc/cove-audits/blob/master/2024-12-23_Zellic_Cove.pdf))  
  - ✅ **Pashov** – Jan 16 2025 ([report](https://github.com/Storm-Labs-Inc/cove-audits/blob/master/2025-01-16_Pashov_Cove.pdf))  
  - ✅ **Pashov re-audit** – Apr 19 2025 ([report](https://github.com/Storm-Labs-Inc/cove-audits/blob/master/2025-04-19_Pashov_Cove.pdf))

We've also engaged [Josselin Feist](https://x.com/Montyly), former engineering director at Trail of Bits and creator of [slither](https://github.com/crytic/slither), to conduct an end-to-end security review and ensure our invariant coverage is complete:

<img width="602" alt="Image" src="https://github.com/user-attachments/assets/42fd8b40-ad47-40ab-ad5a-2ede268aad78" />

#### Jump in
- **Repo:** `git clone https://github.com/Storm-Labs-Inc/cove-contracts-core`
- **Key contract:** [`BasketToken.sol`](https://github.com/Storm-Labs-Inc/cove-contracts-core/blob/master/src/BasketToken.sol)
- **Docs & RFCs:** [docs.cove.finance](https://docs.cove.finance) + [technical RFC](https://docs.cove.finance/technical/cove/rfc)

Hope others find this useful.

**Comment by ernestognw** [comment]
Sharing an opinionated version of an ERC4626 that doesn't comply with ERC-7540:

https://github.com/OpenZeppelin/openzeppelin-contracts/pull/5710

The implementations are designed to be pull-based, liquid and priced at execution by default. I realized these are the base cases, so for example push based can be added on top quite easily, a stepped redemption can be implemented with a minimum cooldown period or a cliff on top of the deposit/withdraw schedule, and finally the price at execution allows the user to decide when to execute without being forced to do so at a certain price.

I believe those versions are cleaner and should be better in the library than 7540

**Comment by giorgilagidze** [comment]
Hi all, is this not included in OZ yet ? when will it ?

**Comment by 0xCSMNT** [comment]
Hi @ernestognw thanks for your thoughtful comments here. I have just re-read them for a second time and don't really have any specific pushback on your ideas. 

However, IMO there is a lot of convergence around the ERC-7540 standard as is, and the RWA space is starved for a unified way of handling these kind of interactions between protocols. While it is by no means a perfect standard, I think it is good enough. Credit to @hieronx and others for pushing it forward. We and many other teams are already implementing it in production code.

Pragmatically speaking, I can see a huge benefit to encouraging this specific standard to be more widely adopted today, rather than to go back to the drawing board to come up with a different solution. The current fragmentation of designs is a major hindrance to RWA adoption currently - teams handle asynchronous transactions in unpredictable ways, and developing ad hoc solutions to adapt them is slow and expensive.

**Comment by hieronx** [comment]
@ernestognw wanted to respond here with some thoughts from my side as well.

I completely agree with a lot of your thoughts on the limitations in the standard. The reality is, this is the outcome of almost 1.5 years (discussions started mid 23, finalized in Oct 24) of consensus finding between many involved parties, including the ERC4626 alliance team, Joey Santoro as one of the original ERC-4626 authors, Superform, Maple Finance, Centrifuge, and many others. Personally I also think there is many parts of the standard that are suboptimal. But, what ultimately matters is that a single standard gets adopted by the market, to be able to use the composability in DeFi and create more value overall.

From that perspective, ERC-7540 has already seen significant success. To give some examples:
- USDai: stablecoin protocol backed by AI infrastructure assets, using ERC7540 for redemptions, with over $580M in TVL
- Centrifuge: tokenization infrastructure with over $1.3B in TVL, using ERC7540 for investments and redemptions
- Sky (Maker) + Grove/Spark (SubDAOs) have integrated ERC7540 for direct investments into RWAs, with currently over $1b in allocations
- Nashpoint: RWA portfolio protocol, using ERC7540 to integrate and allocate to RWAs
- Cove: onchain yield strategies protocol using ERC7540 for redemptions
- Superform: onchain yield management protocol using ERC7540 for RWA allocations

With the increase in tokenization of RWAs and institutions coming into the space, I think now is the time to take the next step and work on increasing composability for these assets. Having an OpenZeppelin implementation is critical to this.

In terms of what can be a base standard, almost all live implementations today follow the same pattern, similar to what I proposed above. Happy to spend more time on this and figure out what is the most valuable base system that can be used by the majority of assets.

**Comment by ernestognw** [comment]
Hey @0xCSMNT and @hieronx, thanks for your comments!

> IMO there is a lot of convergence around the ERC-7540 standard as is, and the RWA space is starved for a unified way of handling these kind of interactions between protocols.

Right, I can attest there's convergence around ERC-7540. 

Given my previous comments, I would like to understand what are the ecosystem-wide benefits of publishing 7540 in OpenZeppelin Contracts even with the limitations I described on @hieronx PR (https://github.com/OpenZeppelin/openzeppelin-contracts/pull/5294). Would you mind sharing what the main advantages of using ERC-7540 are over a customized async ERC-4626 vault? 

My feeling is that ERC7540 makes it difficult to become a standard, for which I would think that a client/frontend/UI could expect to use the same functions behavior as in a regular ERC4626 contract, plus the extended functionality of an ERC7540 vault. Otherwise, I don't see much benefit in treating ERC7540 as a standardized contract.

If we agree it's not perfect, and even the authors stated that [it's not meant to be an OpenZeppelin base contract](https://ethereum-magicians.org/t/eip-7540-asynchronous-erc-4626-tokenized-vaults/16153/9?u=ernestognw), I would say it's worth considering an alternative rather than just following what the ERC describes. It's possible that we put it out, and the fragmentation is still around, perhaps even worse if the standard doesn't fulfill their expectations.

> what ultimately matters is that a single standard gets adopted by the market, to be able to use the composability in DeFi and create more value overall.

I agree with this. The issue is that, although there are successful examples of ERC-7540 implementations, there's still fragmentation around implementations and I'm not sure that publishing an ERC-7540 version in OpenZeppelin Contracts would fix that. Especially if the contract can't be extended easily without inheritance footguns (e.g. knowing where to call super or what to override, see [my comment in the deposit function](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/5294/files#r2114639842), for example)

Even if the standard weren't perfect, the team would be fine adding it to the library but I'm afraid it breaks some extensibility mechanisms that are critical to how the library ensures safety (e.g. with clear instructions about what to override and always making sure of calling `super` to avoid missing side effects).

In this sense, my proposal with https://github.com/OpenZeppelin/openzeppelin-contracts/pull/5710 attempts to be a version that's less invasive to ERC4626 and that maintains inheritance safety. It strips a lot of ERC7540 functionality like the operator or the `pendingDepositRequest`/`pendingWithdrawRequest` functions, but those can still be added on top. We can make it as similar as possible to 7540 if you guys think the other functions are critical to reduce ecosystem fragmentation

**Comment by ernestognw** [comment]
Hi everyone!

We're already working on an ERC-7540 implementation on #6399 targeting the 5.7 release planned for late Q2/early Q3.

Along with it. We're also tracking extension implementations in #6445, #6446 and #6447

Thanks all for your input and patience 💪

**Comment by ernestognw** [comment]
Update for everyone tracking this: an implementation of ERC-7540 has just been merged into [openzeppelin-community-contracts#228](https://github.com/OpenZeppelin/openzeppelin-community-contracts/pull/228) after months of design discussions.

The implementation provides a modular `ERC7540` base plus composable strategy extensions — admin-controlled, time-delayed, and synchronous variants — that can be mixed on the deposit and redeem sides to cover use cases like RWA vaults, liquid staking, and time-locked deposits. Documentation is available [here](https://github.com/OpenZeppelin/openzeppelin-community-contracts/blob/master/docs/modules/ROOT/pages/erc7540.adoc).

Note that the contracts live in `openzeppelin-community-contracts` for now and have not been migrated to the main `openzeppelin-contracts` repository — that migration is pending an audit. Feedback and integrations are very welcome in the meantime.

An epoch-based variant is next on the roadmap.
